src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,529 @@
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/nodes.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/pg_list.h"
+#include "nodes/primnodes.h"
+#if PG_VERSION_NUM >= PG_VERSION_12
+#include "nodes/pathnodes.h"
+#else
+#include "nodes/relation.h"
+#endif
+#include "utils/builtins.h"
+#include "utils/guc.h"
+#include "utils/lsyscache.h"
+
+
+/*
+ * Managed via a GUC
+ */
+int LocalTableJoinPolicy = LOCAL_JOIN_POLICY_AUTO;
+
+typedef struct RangeTableEntryDetails
+{
+	RangeTblEntry *rangeTableEntry;
+	int rteIdentity;
+	List *requiredAttributeNumbers;
+	bool hasConstantFilterOnUniqueColumn;
+} RangeTableEntryDetails;
+
+typedef struct ConversionCandidates
+{
+	List *distributedTableList; /* reference or distributed table */
+	List *localTableList; /* local or citus local table */
+}ConversionCandidates;
+
+typedef struct IndexColumns
+{
+	List *indexColumnNos;
+}IndexColumns;
+
+static bool HasConstantFilterOnUniqueColumn(RangeTblEntry *rangeTableEntry,
+											RelationRestriction *relationRestriction);
+static List * RequiredAttrNumbersForRelation(RangeTblEntry *relationRte,
+											 PlannerRestrictionContext *
+											 plannerRestrictionContext);
+static ConversionCandidates * CreateConversionCandidates(PlannerRestrictionContext *
+														 plannerRestrictionContext,
+														 List *rangeTableList,
+														 Oid resultRelationId);
+static void GetAllUniqueIndexes(Form_pg_index indexForm, List **uniqueIndexes);
+static RangeTableEntryDetails * GetNextRTEToConvertToSubquery(ConversionCandidates *
+															  conversionCandidates,
+															  PlannerRestrictionContext *
+															  plannerRestrictionContext);
+static void RemoveFromConversionCandidates(ConversionCandidates *conversionCandidates,
+										   int rteIdentity);
+static bool AllRangeTableEntriesHaveUniqueIndex(List *rangeTableEntryDetailsList);
+static bool FirstIntListContainsSecondIntList(List *firstIntList, List *secondIntList);
+
+/*
+ * RecursivelyPlanLocalTableJoins gets a query and the planner
+ * restrictions. As long as the query is not plannable by router planner,
+ * it converts either a local or distributed table to a subquery.
+ */
+void
+RecursivelyPlanLocalTableJoins(Query *query,
+							   RecursivePlanningContext *context, List *rangeTableList)
+{
+	PlannerRestrictionContext *plannerRestrictionContext =
+		GetPlannerRestrictionContext(context);
+
+	Oid resultRelationId = InvalidOid;
+	if (IsModifyCommand(query))
+	{
+		resultRelationId = ModifyQueryResultRelationId(query);
+	}
+	ConversionCandidates *conversionCandidates =
+		CreateConversionCandidates(plannerRestrictionContext,
+								   rangeTableList, resultRelationId);
+
+	while (ShouldConvertLocalTableJoinsToSubqueries(query, rangeTableList,
+													plannerRestrictionContext))
+	{
+		RangeTableEntryDetails *rangeTableEntryDetails =
+			GetNextRTEToConvertToSubquery(conversionCandidates,
+										  plannerRestrictionContext);
```

Comment:
One thing I don't quite understand about this algorithm:

We have a query with local tables and distributed tables, which we cannot do directly. My first intuition would be that we can make this query plannable by either recursively planning all the local tables, or recursively planning the distributed tables, and we could use some additional information to decide which we pick (e.g. how many of each? are there filters on unique indexes?)

but it seems like we try to re-decide on every iteration. What I'm trying to understand is: under which circumstances is deciding in each iteration better than deciding once?